### PR TITLE
Implement Webvtt support for HLS

### PR DESCRIPTION
**Pull request by ferrytanu**
Addded WebVttChunkSource which can be used with ChunkSampleSource to
manage downloading subtitle media playlist, and subtitle chunk. For live
streaming subtitle, it refreshes the media playlist every segment
duration, and half of segment duration when the playlist is stale.

Added TimestampOffsetListener file to synch between Audio/Video frame
with subtitle. The value is used as an offset to subtitle timestamp,
this is required esp in live streaming subtitle.

Bringing back X-TIMESTAMP-MAP to handle VOD with initial PTS > 0

Issue: #151

### CODE REVIEW

**Comment by ojw28** [comment]
[Ignore this comment. Just trying to kick googlebot] I signed it!

**Comment by ojw28** [comment]
Note that it's necessary to sign the Google CLA to contribute to this repository, as described [here](https://github.com/google/ExoPlayer/blob/master/CONTRIBUTING.md). I'll take a look at this this week. From a brief initial scan, I'm not sure I like the extent to which timestamps are being propagated between various components. However it may be possible to clean it up without too much effort.

File: demo/src/main/java/com/google/android/exoplayer/demo/Samples.java [hunk 1]
```diff
@@ -125,6 +125,8 @@ public Sample(String name, String contentId, String uri, int type) {
         + "prog_index.m3u8", PlayerActivity.TYPE_HLS),
     new Sample("Apple ID3 metadata", "http://devimages.apple.com/samplecode/adDemo/ad.m3u8",
         PlayerActivity.TYPE_HLS),
+    new Sample("Apple WebVtt", "https://devimages.apple.com.edgekey.net/streaming/examples/bipbop_16x9/bipbop_16x9_variant.m3u8",
```
Comment (ojw28) [inline demo/src/main/java/com/google/android/exoplayer/demo/Samples.java]: This exact example is already added at L117, so you can revert changes to this file.

File: library/src/main/java/com/google/android/exoplayer/hls/WebvttChunkSource.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+            for(Subtitle subtitle: subtitles) {
+                Format format = new Format(subtitle.name, MimeTypes.TEXT_VTT, MediaFormat.NO_VALUE,
+                        MediaFormat.NO_VALUE, MediaFormat.NO_VALUE, MediaFormat.NO_VALUE,
+                        MediaFormat.NO_VALUE, MediaFormat.NO_VALUE, subtitle.language);
+
+                // timestampOffset value is unknown at this time, set 0 as default
+                // subsampleOffset must be set to 0, if not it will be set to OFFSET_SAMPLE_RELATIVE
+                // which will treated as relative time, as a result it won't work for WebVtt
+                // see SubtitleParserHelper for details
+                MediaFormat trackFormat = MediaFormat.createTextFormat(tracks.size(), format.mimeType,
```
Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer/hls/WebvttChunkSource.java]: First argument to createTextFormat should be format.id

File: library/src/main/java/com/google/android/exoplayer/text/SubtitleParserHelper.java [hunk 3]
```diff
@@ -179,8 +179,13 @@ private void handleSample(long sampleTimeUs, SampleHolder holder) {
       if (sampleHolder != holder) {
         // A flush has occurred since this holder was posted. Do nothing.
       } else {
-        this.result = new PlayableSubtitle(parsedSubtitle, subtitlesAreRelative, sampleTimeUs,
-            subtitleOffsetUs);
+        // it's more accurate to read startTime from eventTime compare to sampleTime
```
Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer/text/SubtitleParserHelper.java]: I'm pretty sure you shouldn't need to be changing this class. The sample works perfectly without this. Can we revert it?

Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer/text/SubtitleParserHelper.java]: Oh right, I see what this is for. Although maybe we could clean this up a bit by doing the following:

- If there aren't any events in the subtitle, this.result may as well just be set to null, I think?
- There's no need for the PlayableSubtitle constructor to take startTimeUs. It should be able to calculate it internally.

Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer/text/SubtitleParserHelper.java]: Actually, you still need startTimeUs for the absolute timestamp case. But you could do the following here:

```
        this.result = parsedSubtitle.getEventTimeCount() == 0 ? null : new PlayableSubtitle(
            parsedSubtitle, subtitlesAreRelative, sampleTimeUs, subtitleOffsetUs);
```

And then implement the PlayableSubtitle constructor as:

```
    this.subtitle = subtitle;
    this.offsetUs = (isRelative ? startTimeUs : 0) + offsetUs;
    this.firstEventTimeUs = getEventTime(0);
```

Where `firstEventTimeUs` replaces `startTimeUs` as the member variable.

File: demo/src/main/java/com/google/android/exoplayer/demo/Samples.java (hunk 1 above)
Comment (ferrytanu) [inline demo/src/main/java/com/google/android/exoplayer/demo/Samples.java]: @ojw28 thank you for looking into this. CLA has been signed.

You are right about duplicate sample. I'll remove it.

File: library/src/main/java/com/google/android/exoplayer/text/SubtitleParserHelper.java (hunk 3 above)
Comment (ferrytanu) [inline library/src/main/java/com/google/android/exoplayer/text/SubtitleParserHelper.java]: Gotcha, let's revisit live subtitles on different PR. I'll revert the class for the time being

**Comment by ojw28** [comment]
One other thing to note with this change is that I don't think it'll handle either TS timestamp rollover or discontinuities properly (and it seems very non-trivial to do so, also). Any idea how that might be done within this model?

**Comment by ferrytanu** [comment]
The changes covers that actually.
In WebvttParser, we parse the X-TIMESTAMP-MAP=LOCAL, to get the timestamp, and use that to offset the cue timestamp. As such, the cue timestamp will not be greater than 26.5 hours (which is the limitation of the 33 bits PTS). It serves well so far.

**Comment by ojw28** [comment]
I don't think that's the correct behaviour, unfortunately. Other HLS components use PtsTimestampAdjuster to un-wrap the timestamps. The timestamps parsed from X-TIMESTAMP-MAP are not subject to this un-wrapping logic, and so do wrap. So you'll end up with the WebVTT timestamps wrapping but the timestamps for other media streams not wrapping, I think?

I have a feeling it'll fall out nicer if you parse just X-TIMESTAMP-MAP from the sample somewhere in the hls package, where the timestamp adjustment can be applied.

**Comment by ojw28** [comment]
Discontinuities are even more problematic. HlsChunkSource necessarily instantiates a new PtsTimestampAdjuster when it sees a discontinuity. Matching discontinuities should be present in the WebVTT playlist too, but it's necessary to make sure the right PtsTimestampAdjuster is applied before and after each discontinuity, which is not currently enforced.

**Comment by ojw28** [comment]
I think things actually fall out more nicely using the existing HlsChunkSource implementation in conjuction with a lightweight implementation of the `Extractor` interface specific for HLS+WebVtt.

The `WebVttExtractor` would scan just as much of each sample as is required to establish the TS->LOCAL mapping and the first cue timestamp. This would allow the correct timestamp to be calculated and pushed through `PtsTimestampAdjuster`, which is in-line with what `TsExtractor` does for the other media streams, and ensures that rollover is handled the same in both cases. Handling discontinuities would require extending `PtsTimestampAdjuster` to be discontinuity aware, rather than replacing it with a new instance at each discontinuity point as is currently the case.

Note that with the above approach `WebvttParser` does not need to parse `X-TIMESTAMP-MAP`, which would be done by `WebVttExtractor`. This is actually quite a nice property, because it keeps the HLS specific logic isolated within the hls package.

I intend to give this approach a try over the next couple of days, to see if it's viable.

**Comment by ferrytanu** [comment]
Sounds like that's the more right approach. Wondering how we can avoid calling the webvtt parsing twice, as it will be in WebVttExtractor and SubtitleParserHelper.
Thanks for looking into this, can't wait to see how it being refactor-ed. :smile:

**Comment by ojw28** [comment]
We wont really be parsing the VTT twice. The extractor part will only be parsing as far as the first cue header; it wont need to look any further in the data. I have a hacked together version sort of working, and will look more next week.

**Comment by amed500** [comment]
Hey @ojw28  i have just question about this Pull, will it be merged in the new version of ExoPlayer ?, i need WebVtt Subtitle in My application, for now the version 1.5.2 for Hls it shows only Closed captions subtitle. thank you

**Comment by ojw28** [comment]
This change will not be merged, but we are actively working on a different approach to enabling WebVTT + HLS. Some of the first changes were pushed yesterday (https://github.com/google/ExoPlayer/commit/4bb8bea9528a803a201b809e8bd4abe3f0b460e8, https://github.com/google/ExoPlayer/commit/72f093c4f6d66cea021756e1cbe931b52f55a79e) and more will follow shortly.

**Comment by amed500** [comment]
@ojw28  thank you for your fast answer, will be waiting this feature then, but the multi Audio language Hls will be supported too ?

**Comment by ojw28** [comment]
It's a little further out. There's only so much we can do at once ;). No ETA as of this moment.

**Comment by amed500** [comment]
@ojw28  i really need that feature also, is there any work around you could suggest me to implement this feature ? i saw some workings from this issue #514, thanks

**Comment by tommymc** [comment]
@ojw28 i've seen, you've written WebVttExtractor in 1.5.3rc. Can you add add sample how we can use webvtt + hls? thx

**Comment by ojw28** [comment]
It's not yet possible to fully wire everything up. I guess it's a little misleading that the class was part of the release, given it's still work in progress. Sorry about that!

**Comment by tommymc** [comment]
@ojw28 the version of @ferrytanu use webvtt internal, i means a lots of small webvtt files in playlist m3u8, does your refactor version support webvtt extenal (only one vtt file for one language of the video)?
Thx

**Comment by ojw28** [comment]
Closing this because support should be landing next week. Much thanks for the effort though; it was definitely a helpful reference!

**Pull request closed without merging**